design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,367 @@
+| 1    | Read all K8s Secrets that are marked with labels and update them with Conjur secrets |                                                              | - Less intervention on our part as a customer can define which types of secrets we should parse<br /><br />- K8s native, as-is solution | - Not generalized so we are still K8s bound |
+| 2    | Get secrets by configuration in Secrets Provider             | Define K8S_SECRETS in `cyberark-secrets-provider` manifest<br />` name: K8S_SECRET` |                                                              | - Repetitive work<br /><br />- Not scalable |
+
+To know which K8s Secrets the apps in the namespace need we believe the first solution, Read all K8s Secrets that are marked with labels, would be the most K8s native and painless solution for the customer. Also it avoids the repetitive work of having to list all the K8s Secrets each app needs.
+
+It is important to note however, that we are responsible for providing a solution for updating the K8s secrets and *not* how those Pods will get those secrets.
+
+##### **Customer Experience**
+
+K8s offers the option of adding labels on K8s secrets. For example, a customer would add a label on a K8s Secret like so:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-secret
+  labels:
+    key: CONJUR <----
+  type: Opaque
+stringData:
+  conjur-map: |-
+    username: secrets/my_username
+```
+
+These labels do not have to be Conjur-specific and the customer would decide which label they would like to put in the K8s secrets. They would define which secrets to iterate over by "K8S_SECRET_LABEL" env variable in their Secrets Provider manifest. The Secrets Provider will perform a search for secrets with that label against the [K8s](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#secret-v1-core)/[OC](https://docs.openshift.com/container-platform/4.4/rest_api/index.html#secret-v1-core) API and search for the "conjur_map" on the results.
+
+Note that just because the customer added a label for us to filter over doesn't mean that all secret entries in the K8s Secret are stored in Conjur but it gives us a way to perform a more focused search for "conjur-map" in the K8s Secrets.
+
+Secrets Provider Manifest with K8S_SECRET_LABEL environment variable
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: secrets-provider-deployment
+spec:
+  replicas: 1
+  template:
+    spec:
+      serviceAccountName: sa-1
+      containers:
+      - image: secrets-provider:latest
+        name: cyberark-secrets-provider
+   ...
+      env:
+        - name: K8S_SECRET_LABEL
+          value: CONJUR,PROD
+```
+
+These labels will have an OR relationship not an AND relationship. In other words, a K8s Secret can either have a CONJUR label or PROD label.
```

Comment:
Why you choose OR ? seems AND gives better experience, no?

Reply:
I disagree because a customer may use different labels in K8s Secrets that hold Conjur paths. With an OR relationship, we allow them to list any Label grouping. If we enforce an AND relationship, we would only be pulling secrets that have the label combination listed which seems to be very limiting